pyiron_atomistics/vasp/base.py

```diff
@@ -1843,15 +1909,70 @@ def __init__(self):
         self.incar = Incar(table_name="incar")
         self.kpoints = Kpoints(table_name="kpoints")
         self.potcar = Potcar(table_name="potcar")
-
+        self.options = VaspSpecificOptions()
         self._eddrmm = "warn"
 
+        self.structure = None
+        self.__idx_user_to_pyiron = []
+        self.__idx_pyiron_to_user = []
+
+    @property
+    def _idx_user_to_pyiron(self):
+        if self.structure is None:
+            return None
+        if len(self.__idx_user_to_pyiron) == 0:
+            self._map_pyiron_to_user_idx()
+        return self.__idx_user_to_pyiron
+
+    @property
+    def _idx_pyiron_to_user(self):
+        if self.structure is None:
+            return None
+        if len(self.__idx_pyiron_to_user) == 0:
+            self._map_pyiron_to_user_idx()
+        return self.__idx_pyiron_to_user
+
+    def _map_pyiron_to_user_idx(self):
+        """
+        This writes the indices maps for user->pyiron (_idx_user_to_pyiron), and pyiron->user (_idx_pyiron_to_user)
+        This looks in Input.options for the allow_structure_reordering boolean value, and decides to return either
+        1. Sorting map when species-based reordering is allowed (default behaviour of pyiron)
+        or
+        2. Sorting map when sorting is forbidden, in which case naive map (1:1, 2:2 etc.) is returned.
+        """
+        if self.options.allow_structure_reordering:
+            atom_numbers = self.structure.get_number_species_atoms()
+
+            _idx_user_to_pyiron = list()
+            for species in atom_numbers.keys():
+                indices = self.structure.select_index(species)
+                for i in indices:
+                    _idx_user_to_pyiron.append(i)
+            self.__idx_user_to_pyiron = np.array(_idx_user_to_pyiron)
+
+            _idx_pyiron_to_user = np.array([0] * len(_idx_user_to_pyiron))
+            for i, p in enumerate(_idx_user_to_pyiron):
+                _idx_pyiron_to_user[p] = i
```

Comment:
This can be done without the for-loop

```suggestion
            n = len(_idx_user_to_pyiron)
            _idx_pyiron_to_user = np.empty(n, dtype=int)
            _idx_pyiron_to_user[_idx_user_to_pyiron] = np.arange(n)
```

I don't think we need a `dtype=int` in the arange, but maybe double check `_idx_pyiron_to_user` is still int-like.